File: src/dlc/bsearch.py
from typing import List

def search(nums: List[int], target: int) -> int:
    """704. Binary Search"""
    
    l = 0
    r = len(nums)-1

    while r-l > 1:
        m = (l+r)//2
        if nums[m] == target:
            return m
        elif nums[m] > target:
            r = m
        else:           # nums[m] > target
            l = m

    if nums[l] == target:
        return l
    elif nums[r] == target:
        return r
    else:
        return -1

File: cpp/src/bsearch.cpp
#include "dlc_core/array.h"
#include <vector>

int search(std::vector<int>& nums, int target) {    // 704. Binary Search
    int l = 0;
    int r = nums.size() - 1;

    while (r-l > 1) {
        int m = l + (r-l)/2;
        if (nums[m] == target) {
            return m;
        } else if (nums[m] > target) {
            r = m;
        } else {    // nums[m] < target
            l = m;
        }
    }
    if (nums[l] == target) {
        return l;
    } else if (nums[r] == target) {
        return r;
    } else {
        return -1;
    }
}